hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/source/enumerator/AbstractHoodieEnumerator.java

```diff
@@ -0,0 +1,182 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.source.enumerator;
+
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.source.split.HoodieSourceSplit;
+import org.apache.hudi.source.split.SplitProvider;
+import org.apache.hudi.source.split.SplitRequestEvent;
+
+import org.apache.flink.api.connector.source.SourceEvent;
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.connector.source.SupportsHandleExecutionAttemptSourceEvent;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Abstract class for Hoodie SourceSplit enumerator.
+ */
+abstract class AbstractHoodieEnumerator
+    implements SplitEnumerator<HoodieSourceSplit, HoodieEnumeratorState>,
+    SupportsHandleExecutionAttemptSourceEvent {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractHoodieEnumerator.class);
+
+  private final SplitEnumeratorContext<HoodieSourceSplit> enumeratorContext;
+  private final Map<Integer, String> readersAwaitingSplit;
+  private final SplitProvider splitProvider;
+  private final AtomicReference<CompletableFuture<Void>> availableFuture;
+
+  AbstractHoodieEnumerator(
+      SplitEnumeratorContext<HoodieSourceSplit> enumeratorContext,
+      SplitProvider splitProvider) {
+    this.enumeratorContext = enumeratorContext;
+    this.readersAwaitingSplit = new LinkedHashMap<>();
+    this.availableFuture = new AtomicReference<>();
+    this.splitProvider = splitProvider;
+    this.enumeratorContext
+        .metricGroup()
+        // This number may not capture the entire backlog due to split discovery throttling to avoid
+        // excessive memory footprint. Some pending splits may not have been discovered yet.
+        .setUnassignedSplitsGauge(() -> Long.valueOf(splitProvider.pendingSplitCount()));
+  }
+
+  @Override
+  public void start() {
+    splitProvider.start();
+  }
+
+  @Override
+  public void handleSplitRequest(int subtaskId, @Nullable String requesterHostname) {
+    readersAwaitingSplit.put(subtaskId, requesterHostname);
+    assignSplits();
+  }
+
+  @Override
+  public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
+    if (sourceEvent instanceof SplitRequestEvent) {
+      SplitRequestEvent splitRequestEvent = (SplitRequestEvent) sourceEvent;
```

Comment:
wondering what's the difference between `SplitRequestEvent` and `handleSplitRequest(int subtaskId, @Nullable String requesterHostname)`, looks like this method is be invoked after a successful split handling from a task.

Reply:
handleSourceEvent is mainly for customized protocol between the reader and the enumerator. We piggyback finished split id from reader back to enumerator. It will help us to emit metrics in the future.